File: leetcode/contest-247/1916. Count Ways to Build Rooms in an Ant Colony.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1e5 + 10, M = N;
const int P = 1e9 + 7;
typedef long long LL;

int head[N], ver[M], Next[M], tot;
int f[N], g[N];
void add(int x, int y) {
    ver[++tot] = y;
    Next[tot] = head[x], head[x] = tot;
}
int qpow(int a, int b, int p) {
    int ans = 1;
    while (b) {
        if (b & 1) ans = (LL)ans * a % p;
        a = (LL)a * a % p;
        b >>= 1;
    }
    return ans;
}
int s[N], sz[N];
int dfs(int u) {
    sz[u] = 0;
    for (int i = head[u]; i; i = Next[i]) {
        int y = ver[i];
        dfs(y);
        sz[u] += sz[y];
    }
    s[u] = f[sz[u]];
    for (int i = head[u]; i; i = Next[i]) {
        int y = ver[i];
        s[u] = (LL)s[u] * g[sz[y]] % P * s[y] % P;
    }
    sz[u]++;
    return s[u];
}
class Solution {
   public:
    int waysToBuildRooms(vector<int>& prevRoom) {
        int n = prevRoom.size();
        tot = 0;
        // 点的标号为 [0, n-1]
        for (int i = 0; i < n; i++) {
            head[i] = 0;
        }
        for (int i = 1; i < n; i++) {
            add(prevRoom[i], i);
        }
        f[0] = g[0] = 1;
        for (int i = 1; i <= n; i++) {
            f[i] = (LL)f[i - 1] * i % P;
            g[i] = (LL)g[i - 1] * qpow(i, P - 2, P) % P;
        }
        return dfs(0);
    }
};

File: leetcode/残酷刷题/1916. Count Ways to Build Rooms in an Ant Colony.ts
// function pow (a: number, b: number, p: number) {
//   let ans = 1 % p
//   while (b) {
//     if (b & 1) ans = Number(BigInt(ans) * BigInt(a) % BigInt(p))
//     a = Number(BigInt(a) * BigInt(a) % BigInt(p))
//     b >>= 1
//   }
//   return ans
// }
function pow (a: bigint, b: bigint, p: bigint) {
  let ans = 1n % p
  while (b) {
    if (b & 1n) ans = ans * a % p
    a = a * a % p
    b >>= 1n
  }
  return ans
}
function waysToBuildRooms (prevRoom: number[]): number {
  const n = prevRoom.length
  const e: number[][] = [...new Array(n)].map(() => [])
  const size: number[] = new Array(n).fill(0)
  const f: Array<bigint> = new Array(n).fill(0n)
  const g: Array<bigint> = new Array(n).fill(0n)
  const p = BigInt(1e9 + 7)

  f[0] = g[0] = 1n
  for (let i = 1; i < n; i++) {
    f[i] = f[i - 1] * BigInt(i) % p
    g[i] = g[i - 1] * pow(BigInt(i), p - 2n, p) % p
  }

  for (let i = 1; i < n; i++) {
    e[prevRoom[i]].push(i)
  }
  return Number(dfs(0))

  function dfs (x: number) {
    let ans = 1n
    for (const y of e[x]) {
      const cnt = dfs(y)
      size[x] += size[y]
      ans = (ans * cnt % p) * g[size[y]] % p
    }
    ans = ans * f[size[x]] % p
    size[x] += 1
    return ans % p
  }
};
